protocol/encoder/encoder.go

```diff
@@ -0,0 +1,82 @@
+package encoder
+
+import (
+	"encoding/binary"
+)
+
+type Encoder struct {
+	bytes  []byte
```

Comment:
I think you can use bytes.Buffer here and not have to track offset/bytes etc. on your own: https://pkg.go.dev/bytes#Buffer